feat: Add liveReload (#1407)
SAP/ui5-tooling: internal/documentation/docs/pages/Server.md

This middleware lists project files with URLs under several `/discovery` endpoints. This is exclusively used by the OpenUI5 test suite application.

### liveReload

When live reload is active, the UI5 server opens a WebSocket connection that notifies the browser of source changes and triggers a page reload.

> Reloads are driven by the `BuildServer`, which emits a debounced `sourcesChanged` event whenever watched source files change. A burst of changes therefore results in a single reload notification.

Reviewer: I really like the detailed documentation but I would recommend to have a more abstract / consumer friendly first section and then describing the technical details in a second section. The consumer has to know whats the benefit of this middleware and how to control it. IMHO The underlying technical aspects are only relevant for experts or custom middleware developers.

Reviewer 2: Yes, that makes sense.

Author: Totally agree. There are now two sections.

> Reloads are driven by the `BuildServer`, which emits a debounced `sourcesChanged` event (100ms debounce) whenever watched source files change. A burst of changes therefore results in a single reload notification.

Reviewer 3: ```suggestion
Reloads are driven by the `BuildServer`, which emits a debounced `sourcesChanged` event whenever watched source files change. A burst of changes therefore results in a single reload notification.
```

Maybe don't mention the exact number since that might change in the implementation and is not really relevant to the user.

The `serveLiveReloadClient` middleware serves the client script at `/.ui5/liveReload/client.js`. The script tag is automatically injected into HTML responses by the `serveResources` middleware.

> To prevent intermediate proxies from idle-closing the WebSocket, the client sends a `{type: "ping"}` keepalive every 30 seconds while the connection is open. The server echoes the same message back.

Reviewer 3: ```suggestion
To prevent intermediate proxies from idle-closing the WebSocket, the client sends a keepalive message every 30 seconds while the connection is open. The server echoes the same message back.
```

Like above, the message content is an implementation detail that is not relevant in the user facing documentation

This middleware lists project files with URLs under several `/discovery` endpoints. This is exclusively used by the OpenUI5 test suite application.

### liveReload

Live reload automatically refreshes the browser whenever you change a source file in your project — no manual reload needed. This shortens the edit/test cycle during development.

#### Usage

Live reload is **enabled by default** with `ui5 serve`. Open your app in the browser, edit a source file, save — the page reloads.

To control it:

- **CLI flag**: `ui5 serve --live-reload` / `--no-live-reload`
- **Project configuration** (specVersion 5.0+): `server.settings.liveReload` in `ui5.yaml`. See [Configuration](./Configuration.md).

> When the dev server is restarted, the browser automatically reconnects and reloads once the server is back. Saving multiple files at once triggers a single reload, not one per file.

Reviewer: ```suggestion
When the development server is restarted, the browser automatically reconnects and reloads the browser tab. Saving multiple files at once triggers a single reload, not one per file.
```

#### Technical Details

The following describes how the middleware works internally. This is relevant for advanced users and custom middleware developers — not required for regular usage.

> When live reload is active, the UI5 server opens a WebSocket connection that notifies the browser of source changes and triggers a page reload.

Reviewer: ```suggestion
When live reload is active, the development server opens a WebSocket connection that notifies the browser of source changes and triggers a page reload.
```

> Reloads are driven by the `BuildServer`, which emits a debounced `sourcesChanged` event whenever watched source files change. A burst of changes therefore results in a single reload notification.

Reviewer: ```suggestion
Reloads are driven by the `BuildServer` module, which emits a debounced `sourcesChanged` event whenever watched source files change. A burst of changes therefore results in a single reload notification.
```

The `liveReloadClient` middleware serves the client script at `/.ui5/liveReload/client.js`. The script tag is automatically injected into HTML responses by the `serveResources` middleware.

To prevent intermediate proxies from idle-closing the WebSocket, the client sends a keepalive message every 30 seconds while the connection is open. The server echoes the same message back.

> When the WebSocket connection is lost (e.g. because the server was restarted), the client polls the WebSocket endpoint every second and reloads the page once the server accepts connections again. While the browser tab is hidden, polling pauses until it becomes visible again.

Reviewer: ```suggestion
When the WebSocket connection is lost (e.g. because the development server was restarted), the client polls the WebSocket endpoint every second and reloads the page once the server accepts connections again. While the browser tab is hidden, polling pauses until it becomes visible again.
```
